src/Utilities/AnsiBuilder.cs

```diff
@@ -0,0 +1,330 @@
+                    isLeadingZero = false;
+
+                // Write the last digit (unless a leading zero)
+                if (!isLeadingZero)
+                    buffer[bufferPos + nbChars++] = (char) ('0' + valueLong % 10);
+
+                // Add the decimal point
+                if (--nbDecimals == 0 && !isLeadingZero)
+                    buffer[bufferPos + nbChars++] = '.';
+
+                valueLong /= 10;
+            }
+
+            bufferPos += nbChars;
+
+            // Reverse the result
+            for (var i = nbChars / 2 - 1; i >= 0; i--)
+            {
+                var c = buffer[bufferPos - i - 1];
+                buffer[bufferPos - i - 1] = buffer[bufferPos - nbChars + i];
+                buffer[bufferPos - nbChars + i] = c;
+            }
+
+            return this;
+        }
+
+        /// <summary>
+        /// Replace all occurrences of a string by another one
+        /// </summary>
+        public AnsiBuilder Replace(string oldStr, string newStr)
+        {
+            if (bufferPos == 0)
+                return this;
+
+            replacement ??= new List<char>();
+
+            // Create the new string into replacement
+            for (var i = 0; i < bufferPos; i++)
+            {
+                var isToReplace = false;
+                // If first character found, check for the rest of the string to replace
+                if (buffer[i] == oldStr[0])
+                {
+                    var k = 1;
+                    while (k < oldStr.Length && buffer[i + k] == oldStr[k])
+                        k++;
+                    isToReplace = k >= oldStr.Length;
+                }
+
+                if (isToReplace)
+                {
+                    i += oldStr.Length - 1;
+                    if (newStr == null) continue;
+                    foreach (var t in newStr)
+                        replacement.Add(t);
+                }
+                else
+                {
+                    replacement.Add(buffer[i]);
+                }
+            }
+
+            // Copy back the new string into chars
+            ReallocateIfn(replacement.Count - bufferPos);
+            for (var k = 0; k < replacement.Count; k++)
+                buffer[k] = replacement[k];
+            bufferPos = replacement.Count;
+            replacement.Clear();
+            return this;
+        }
+
+        private void ReallocateIfn(int nbCharsToAdd)
+        {
+            if (bufferPos + nbCharsToAdd <= charsCapacity) return;
+            charsCapacity = Math.Max(charsCapacity + nbCharsToAdd, charsCapacity * 2);
+            var newChars = new char[charsCapacity];
+            buffer.CopyTo(newChars, 0);
+            buffer = newChars;
+        }
+
+        /// <summary>Parses a string for special tags and replaces them with ANSI codes.</summary>
+        private void Parse()
+        {
+            // Seek out all tags and replace them with the equivalent ANSI codes.
```

Comment:
Although this whole token replacement strategy had a number of issues, I detailed and addressed some of them at my own PR #108...  Further, another dev has just come along and wants to work with DataFormatter (#106).  If we're all going to happen to pick overlapping areas, we're going to need to coordinate closer to avoid overlapped efforts.  :)